service/history/historyEngine.go

```diff
@@ -1767,18 +1769,10 @@ func (e *historyEngineImpl) RequestCancelWorkflowExecution(
 				}
 			}
 
-			isCancelRequested, cancelRequestID := mutableState.IsCancelRequested()
+			isCancelRequested := mutableState.IsCancelRequested()
 			if isCancelRequested {
-				cancelRequest := req.CancelRequest
-				if cancelRequest.GetRequestId() != "" {
-					requestID := cancelRequest.GetRequestId()
-					if requestID != "" && cancelRequestID == requestID {
-						return updateWorkflowWithNewWorkflowTask, nil
-					}
-				}
-				// if we consider workflow cancellation idempotent, then this error is redundant
-				// this error maybe useful if this API is invoked by external, not workflow task from transfer queue.
-				return nil, ErrCancellationAlreadyRequested
+				// since cancellation is idempotent
+				return &updateWorkflowAction{noop: true}, nil
 			}
 
 			if _, err := mutableState.AddWorkflowExecutionCancelRequestedEvent("", req); err != nil {
```

Comment:
Probably not relevant to your change, but I think we should remove cause from `AddWorkflowExecutionCancelRequestedEvent`.  It does not makes sense for requestCancel.

Reply:
Let me make a PR for this, probably tomorrow

Reply:
#957